esp/services/ws_decoupledlogging/ws_decoupledloggingservice.cpp

```diff
@@ -89,64 +82,122 @@ void CWSDecoupledLogEx::init(IPropertyTree* cfg, const char* process, const char
             if(!logThread)
                 throw MakeStringException(-1, "Failed to create update log thread for %s", agentName);
-    
+
             CLogRequestReader* logRequestReader = logThread->getLogRequestReader();
             if (!logRequestReader)
                 throw MakeStringException(-1, "CLogRequestReader not found for %s.", agentName);
-            group->loggingAgentThreads.push_back(logThread);
+
+            if (group->getLoggingAgentThread(agentName))
+                throw MakeStringException(-1, "%s: >1 logging agents are named as %s.", groupName, agentName);
+            group->addLoggingAgentThread(agentName, logThread);
         }
-        logGroups.push_back(group);
+        logGroups.insert({groupName, group});
     }
 }
 
 bool CWSDecoupledLogEx::onGetLogAgentSetting(IEspContext& context, IEspGetLogAgentSettingRequest& req, IEspGetLogAgentSettingResponse& resp)
 {
     try
     {
-        BoolHash logAgentFound;
-        StringBuffer errorStatus;
-        StringArray& agentNames = req.getAgentNames();
-        IArrayOf<IEspLogAgentSetting> logAgentSettings;
-        for (auto gp : logGroups)
+        IArrayOf<IEspLogAgentGroupSetting> groupSettingResp;
+        IArrayOf<IConstLogAgentGroup>& groups = req.getGroups();
+        if (!groups.ordinality())
         {
-            const char* tankFileDir = gp->tankFileDir.get();
-            const char* tankFileMask = gp->tankFileMask.get();
-            for (auto in : gp->loggingAgentThreads)
+            for (auto ml : logGroups)
             {
-                IEspLogAgent* agent = in->getLogAgent();
-                const char* agentName = agent->getName();
-                if (!checkName(agentName, agentNames, true))
-                    continue;
+                IArrayOf<IEspLogAgentSetting> agentSettingResp;
+                getSettingsForAllLoggingAgentsInGroup(ml.second, agentSettingResp);
+
+                Owned<IEspLogAgentGroupSetting> groupSetting = createLogAgentGroupSetting();
+                groupSetting->setGroupName(ml.first.c_str());
+                groupSetting->setGroupStatus("Found");
+                const char* tankFileDir = ml.second->getTankFileDir();
+                const char* tankFileMask = ml.second->getTankFileMask();
+                groupSetting->setTankFileDir(tankFileDir);
+                if (!isEmptyString(tankFileMask))
+                    groupSetting->setTankFileMask(tankFileMask);
+                groupSetting->setLogAgentSetting(agentSettingResp);
+                groupSettingResp.append(*groupSetting.getClear());
+            }
+            resp.setSettings(groupSettingResp);
+            return true;
+        }
+
+        ForEachItemIn(i, groups)
+        {
+            IConstLogAgentGroup& g = groups.item(i);
+            const char* gName = g.getGroupName();
+            if (isEmptyString(gName))
+                throw MakeStringException(ECLWATCH_INVALID_INPUT, "Group name not specified.");
```

Comment:
It **would be nice** if an ordinality of one with no agents or a single empty agent could be recognized as a request for all. Using the method's form interface,  a user could add an array item to see what options are available, add no data, and submit a request that will be rejected.